plugins/nf-google/src/main/nextflow/cloud/google/nio/GsPath.groovy

```diff
@@ -0,0 +1,168 @@
+import nextflow.file.ETagAwareFile
+/**
+ * Implements an ETag-aware wrapper around the NIO filesystem
+ * provided by the Google Cloud SDK.
+ *
+ * @author Ben Sherman <bentshermann@gmail.com>
+ */
+@Slf4j
+@CompileStatic
+class GsPath implements Path, ETagAwareFile {
+
+    @Delegate
+    CloudStoragePath target
+
+    private GsFileSystem fileSystem
+
+    private Blob metadata
+
+    GsPath(CloudStoragePath target) {
+        this(new GsFileSystem(target.fileSystem), target)
+    }
+
+    GsPath(GsFileSystem fileSystem, CloudStoragePath target) {
+        this.target = target
+        this.fileSystem = fileSystem
+    }
+
+    @Override
+    int compareTo(Path other) {
+        return target.compareTo(unwrap0(other))
+    }
+
+    @Override
+    boolean equals(Object other) {
+        return target.equals(other instanceof GsPath ? other.target : other)
+    }
+
+    @Override
+    Path getFileName() {
+        return wrap0(target.getFileName())
+    }
+
+    @Override
+    Path getName(int index) {
+        return wrap0(target.getName(index))
+    }
+
+    @Override
+    Path getParent() {
+        return wrap0(target.getParent())
+    }
+
+    @Override
+    Path getRoot() {
+        return wrap0(target.getRoot())
+    }
+
+    @Override
+    GsFileSystem getFileSystem() {
+        return fileSystem
+    }
+
+    @Override
+    int hashCode() {
+        return target.hashCode()
+    }
+
+    @Override
+    Path normalize() {
+        return wrap0(target.normalize())
+    }
+
+    @Override
+    Path relativize(Path other) {
+        return wrap0(target.relativize(unwrap0(other)))
+    }
+
+    @Override
+    Path resolve(String other) {
+        return wrap0(target.resolve(other))
+    }
+
+    @Override
+    Path resolve(Path other) {
+        return wrap0(target.resolve(unwrap0(other)))
+    }
+
+    @Override
+    Path resolveSibling(String other) {
+        return wrap0(target.resolveSibling(other))
+    }
+
+    @Override
+    Path resolveSibling(Path other) {
+        return wrap0(target.resolveSibling(unwrap0(other)))
+    }
+
+    @Override
+    Path subpath(int beginIndex, int endIndex) {
+        return wrap0(target.subpath(beginIndex, endIndex))
+    }
+
+    @Override
+    Path toAbsolutePath() {
+        return wrap0(target.toAbsolutePath())
+    }
+
+    @Override
+    Path toRealPath(LinkOption... options) {
+        return wrap0(target.toRealPath(options))
+    }
+
+    @Override
+    String toString() {
+        return target.toString()
+    }
+
+    @Override
+    String getETag() {
+        if( metadata == null ) {
+            final googleOpts = GoogleOpts.create(Global.session as Session)
+            final client = StorageOptions.newBuilder()
+                    .setProjectId(googleOpts.projectId)
+                    .build()
+                    .getService()
+            metadata = client.get(target.bucket(), target.toString())
+        }
+        return metadata?.getMd5ToHexString()
+    }
```

Comment:
I added a custom NIO filesystem for Google cloud storage which just delegates to the existing one, but allows us to add custom behavior like this ETag support.

It turns out that the ETag returned by the Java SDK is a CRC32 instead of MD5. The MD5 is also available through this storage API, but I haven't gotten it to work yet. The metadata is always null, even though I provided the projectId and I have the credentials file. Need to investigate further when I have more time